enclave/core/sgx/entropy.c

```diff
@@ -1,48 +1,82 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include <openenclave/bits/safecrt.h>
+#include <openenclave/corelibc/string.h>
 #include <openenclave/enclave.h>
 #include <openenclave/internal/entropy.h>
+#include <openenclave/internal/raise.h>
 #include <openenclave/internal/rdrand.h>
+#include <openenclave/internal/rdseed.h>
+#include "cpuid.h"
 
-/* TODO: This should use RDSEED instead. See issue #242. */
-oe_result_t oe_get_entropy(void* output, size_t len)
+typedef uint64_t (*_entropy_function_t)(void);
+
+OE_INLINE bool _has_cpuid_feature(
+    uint32_t leaf,
+    uint32_t feature,
+    uint32_t feature_register)
 {
-    oe_result_t ret = OE_UNEXPECTED;
-    unsigned char* p = (unsigned char*)output;
+    oe_assert(feature_register < OE_CPUID_REG_COUNT);
+    uint64_t r[OE_CPUID_REG_COUNT] = {0};
+    r[OE_CPUID_RAX] = leaf;
+    return (
+        (oe_emulate_cpuid(
+             &r[OE_CPUID_RAX],
+             &r[OE_CPUID_RBX],
+             &r[OE_CPUID_RCX],
+             &r[OE_CPUID_RDX]) == 0) &&
+        (r[feature_register] & feature));
+}
 
-    if (!output)
-        goto done;
+static oe_entropy_kind_t _get_entropy_kind()
+{
+    oe_entropy_kind_t result = OE_ENTROPY_KIND_NONE;
 
-    /* Copy 64-bit random integers to output */
-    {
-        size_t n = len / sizeof(uint64_t);
+    if (_has_cpuid_feature(1, OE_CPUID_RDRAND_FEATURE, OE_CPUID_RCX))
```

Comment:
What will happen if the host lies about rdrand/rdseed availability? Could it end up being emulated by the hypervisor?

Reply:
Hypervisor doesn't intercept this AFAIK, but malicious host can set arbitrary CPUID values as part of the enclave initialization. The same logic as described in #216, applied to RDRAND instead of AESNI:

- If host lies that RDRAND is absent, enclave initialization stops the enclave from running.
- If host lies that RDRAND is present, it's not a security issue, the enclave simply fails later on first use of RDRAND and fail out with SIGILL.

We do not apply the same requirement to RDSEED because it's less common than RDRAND still and will block simulation mode in many more cases, so the worse lie the host can tell is that RDSEED is absent, but RDRAND is present. This is mitigated by exposing the type of entropy used the caller so that it can either fail out, or enhance the predictive resistance provided by RDRAND themselves as illustrated in this PR.